### ISSUE TITLE
Libtorch JIT :  Does enabling profiling mode increase CPU memory usage ?  How to disable profiling mode properly ?

### ISSUE THREAD

**Issue by w1d2s** [pytorch/pytorch#60433]
Hi, I am trying to deploying an Attention-based Encoder Decoder (AED) model with libtorch C++ frontend, when model's decoder loops at output sequence ( the decoder jit module 's forward method is repeatedly called at each label time step ), the CPU memory usage is very high (~ 20 GB), and I think it's far too high compared to it should be ( at each decoder step, the internal state tensors should occupy about < 400 MB in total, and state tensors at previous steps is released correctly with management of smart pointers).

I call torch::jit::getProfilingMode() at begining of inference, and it's true; I try to set it false, but the memory usage is still high.

I would like to know :
1) whether the high CPU memory usage is related to torch JIT 's profiling mode ?
2) is there any other way to profile CPU memory usage ?

The libtorch version used is 1.9.0

Thanks a lot.
- [Discussion Forum](https://discuss.pytorch.org/)

cc @gmagogsfm

**Comment by nikithamalgifb** [comment]
Hi, thank you for the report! Are you able to share a small reproducer that is seeing this issue? It's difficult to investigate without a repro. Thanks!

cc: @suo @Krovatkin @eellison - Are you guys aware of this? If you aren't the right contact, can you please tag the right person.

**Comment by Krovatkin** [comment]
@w1d2s could you please make a small repo? 

`torch::jit::getProfilingMode() = false;` should disable the profiling executor, so it's probably something else.

1. Could you please profile memory allocations with https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html . Let's see if they come from aten ops or from somewhere else? 
2. @w1d2s is the memory usage **much lower** without using the JIT?

Thanks!

@nikithamalgifb we could keep on the jit queue until we get more info.

**Comment by w1d2s** [comment]
> @w1d2s could you please make a small repo?
> 
> `torch::jit::getProfilingMode() = false;` should disable the profiling executor, so it's probably something else.
> 
> 1. Could you please profile memory allocations with https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html . Let's see if they come from aten ops or from somewhere else?
> 2. @w1d2s is the memory usage **much lower** without using the JIT?
> 
> Thanks!
> 
> @nikithamalgifb we could keep on the jit queue until we get more info.

Thank you for your advice, but I still have a small question: my code is integrated in a beam search decoder written in C++, so should I profile in C++ or I should implement another python version and profile in python? If in C++, how can I configure profiler in C++ to enable CPU memory profiling ?

**Comment by w1d2s** [comment]
> Hi, thank you for the report! Are you able to share a small reproducer that is seeing this issue? It's difficult to investigate without a repro. Thanks!
> 
> cc: @suo @Krovatkin @eellison - Are you guys aware of this? If you aren't the right contact, can you please tag the right person.

Thank you for your reply, I am afraid I cannot post codes here, but I can roughly describe what I do with some example. I am basically implementing an sequence-to-sequence model with standard Transformer, and when decoding I do beam search with Transformer Decoder, and in order to make the decoder inference more efficient, I return states Tensors (at T time step, they are in Size(batch_size, T-1, ...) shape) from this decoder step and pass it to next decoder inference.

my decoder wrapper in C++ is like this:
`

class Decoder {
 public:
  // for Transformer Decoder states.
  using StateType = c10::List<torch::Tensor>;

  // decoder one step inference
  void ForwardOneStepBatch(const std::vector<void*>& y_last_step_vec,
                           void* h,
                           std::vector<void*>* outputs,
                           std::vector<StateType*>& state_vec) {
    // 
    c10::List<torch::Tensor> y_last_step_list;
    c10::List<StateType> state_list;
    
    // group label tensor / state tensors from different beam
    // into one batch, to make inference more efficient.
    for (int i = 0; i < beam_batch_size; ++i) {
      y_last_step_list.push_back(
          torch::from_blob(y_last_step_vec[i], at::IntArrayRef{1}, long_opt));
      state_list.push_back(*(state_vec[i]));
    }

    torch::Tensor encoded = torch::from_blob(
        h, /* shape_of_encoded_vector */, float_opt);
  
    // (time, batch, feat_dim) -> (batch, time, feat_dim)
    torch::Tensor encoded_time_first = encoded.permute({1, 0, 2});

    auto out_tuple = module_
                        ->run_method("step_batch",
                                     y_last_step_list,
                                     encoded_time_first,
                                     state_list)
                        .toTuple()
                        .elements();

    // (T = 1, N, V) -> (N, V) or (V, )
    logit_ = out_tuple[0].toTensor().squeeze().contiguous();
    outputs->clear();
    outputs->push_back(logit_.data_ptr());
    auto new_state_list = out_tuple[2].toList();  // c10::List<IValue>
    // save new state tensors generated at this time step.
    for (int i = 0; i < beam_batch_size_; ++i) {
      (*state_vec[i]) = new_state_list.get(i).to<StateFlat>();
    }
  }

 private:
  torch::NoGradGuard no_grad_;
  torch::Tensor logits_;
  
  std::unique_ptr<torch::jit::module> module_ = nullptr; 

  // some other members ...
};
`
